{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from coretracker.coreutils import Faces\n",
    "from ete3 import Tree, TreeStyle, faces\n",
    "from collections import defaultdict\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-1-238f6d99e65d>, line 16)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-1-238f6d99e65d>\"\u001b[1;36m, line \u001b[1;32m16\u001b[0m\n\u001b[1;33m    def\u001b[0m\n\u001b[1;37m        ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "\n",
    "class CodonStateTransition(object):\n",
    "    def __init__(self, gcode):\n",
    "        pass\n",
    "\n",
    "    def _build_transition(self):\n",
    "        pass\n",
    "    \n",
    "    def get_transition_for(codon):\n",
    "        pass\n",
    "    \n",
    "    def \n",
    "\n",
    "class RNAediting (object):\n",
    "    \"\"\"Attempt to check edition with a simple algorithm\"\"\"\n",
    "    def __init__(self, mtalignment, genome, dt, radius,gcode=1):\n",
    "        self.codontable = CodonTable.unambiguous_dna_by_id[gcode]\n",
    "\n",
    "    def get_aa_consensus(self):\n",
    "        pass\n",
    "\n",
    "    def nuc_changer(self, nuc=\"U\"):\n",
    "        pass\n",
    "        \n",
    "    def nucleotide_scanner(self, nuc=\"U\"):\n",
    "        pass\n",
    "    \n",
    "    def check_editing():\n",
    "        pass\n",
    "\n",
    "    def _cu_subs_checking(self, codon1, amino2):\n",
    "        pass\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['G', 'A', 'B', 'D']\n"
     ]
    }
   ],
   "source": [
    "class Node(object):\n",
    "    \n",
    "    def __init__(self, codon, aa):\n",
    "        self.connected = {}\n",
    "        self.codon = codon\n",
    "        self.aa = aa\n",
    "    \n",
    "    def add_edge(self, node):\n",
    "        if not self.is_connected(node):\n",
    "            # add edge in both direction\n",
    "            self.connected[node.codon] = node\n",
    "            node.connected[self.codon] = self\n",
    "    \n",
    "    def is_connected(self, node):\n",
    "        return node.codon in self.connected.keys()\n",
    "\n",
    "\n",
    "class CodonGraph(object):\n",
    "    \n",
    "    def __init__(self, codon_table):\n",
    "        self.root, self.graph = self._build_graph(codon_table)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import itertools\n",
    "def check_insert(seq, vision, decode_table):\n",
    "    alphabet = ['A', 'U', 'G', 'C']\n",
    "    # possible way to improve\n",
    "    # check end of possible codon for first aa in vision\n",
    "    # and decide which insertion are possible\n",
    "    for pos in range(3):\n",
    "        for size in range(1,4):\n",
    "            for prod in itertools.product(alphabet, repeat=size):\n",
    "                s=seq[:pos]+ \"\".join(prod)+ seq[pos:])[:6]\n",
    "                curaa =  \"\".join(decode_table[s[i:(i+1)*3] for i in len(vision)])\n",
    "                if curaa == vision:\n",
    "                    yield s, prod, pos\n",
    "\n",
    "def check_delete_till(seq, vision, decode_table):\n",
    "    for pos in range(3):\n",
    "        for j in xrange(1,4):\n",
    "            s =  (seq[:pos] + seq[pos+j:])[:6]\n",
    "            curaa =  \"\".join(decode_table[s[i:(i+1)*3] for i in len(vision)])\n",
    "            if curaa == vision:\n",
    "                yield s, seq[pos:pos+j], pos\n",
    "                \n",
    "\n",
    "def get_transition(codon, aa):\n",
    "    \n",
    "    \n",
    "\n",
    "def check_single_mut(seq, vision, decode_table):\n",
    "    curaa =  \"\".join(decode_table[s[i:(i+1)*3] for i in len(vision)])\n",
    "    if curaa[-1] ==  vision[-1]:\n",
    "        number_of_error = get_transition(seq[0:3], vision[0], decode_table)\n",
    "        \n",
    "    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4096\n",
      "AATGGC\n",
      "UATGGC\n",
      "GATGGC\n",
      "CATGGC\n",
      "AAATGG\n",
      "AUATGG\n",
      "AGATGG\n",
      "ACATGG\n",
      "UAATGG\n",
      "UUATGG\n",
      "UGATGG\n",
      "UCATGG\n",
      "GAATGG\n",
      "GUATGG\n",
      "GGATGG\n",
      "GCATGG\n",
      "CAATGG\n",
      "CUATGG\n",
      "CGATGG\n",
      "CCATGG\n",
      "AAAATG\n",
      "AAUATG\n",
      "AAGATG\n",
      "AACATG\n",
      "AUAATG\n",
      "AUUATG\n",
      "AUGATG\n",
      "AUCATG\n",
      "AGAATG\n",
      "AGUATG\n",
      "AGGATG\n",
      "AGCATG\n",
      "ACAATG\n",
      "ACUATG\n",
      "ACGATG\n",
      "ACCATG\n",
      "UAAATG\n",
      "UAUATG\n",
      "UAGATG\n",
      "UACATG\n",
      "UUAATG\n",
      "UUUATG\n",
      "UUGATG\n",
      "UUCATG\n",
      "UGAATG\n",
      "UGUATG\n",
      "UGGATG\n",
      "UGCATG\n",
      "UCAATG\n",
      "UCUATG\n",
      "UCGATG\n",
      "UCCATG\n",
      "GAAATG\n",
      "GAUATG\n",
      "GAGATG\n",
      "GACATG\n",
      "GUAATG\n",
      "GUUATG\n",
      "GUGATG\n",
      "GUCATG\n",
      "GGAATG\n",
      "GGUATG\n",
      "GGGATG\n",
      "GGCATG\n",
      "GCAATG\n",
      "GCUATG\n",
      "GCGATG\n",
      "GCCATG\n",
      "CAAATG\n",
      "CAUATG\n",
      "CAGATG\n",
      "CACATG\n",
      "CUAATG\n",
      "CUUATG\n",
      "CUGATG\n",
      "CUCATG\n",
      "CGAATG\n",
      "CGUATG\n",
      "CGGATG\n",
      "CGCATG\n",
      "CCAATG\n",
      "CCUATG\n",
      "CCGATG\n",
      "CCCATG\n",
      "AATGGC\n",
      "AUTGGC\n",
      "AGTGGC\n",
      "ACTGGC\n",
      "AAATGG\n",
      "AAUTGG\n",
      "AAGTGG\n",
      "AACTGG\n",
      "AUATGG\n",
      "AUUTGG\n",
      "AUGTGG\n",
      "AUCTGG\n",
      "AGATGG\n",
      "AGUTGG\n",
      "AGGTGG\n",
      "AGCTGG\n",
      "ACATGG\n",
      "ACUTGG\n",
      "ACGTGG\n",
      "ACCTGG\n",
      "AAAATG\n",
      "AAAUTG\n",
      "AAAGTG\n",
      "AAACTG\n",
      "AAUATG\n",
      "AAUUTG\n",
      "AAUGTG\n",
      "AAUCTG\n",
      "AAGATG\n",
      "AAGUTG\n",
      "AAGGTG\n",
      "AAGCTG\n",
      "AACATG\n",
      "AACUTG\n",
      "AACGTG\n",
      "AACCTG\n",
      "AUAATG\n",
      "AUAUTG\n",
      "AUAGTG\n",
      "AUACTG\n",
      "AUUATG\n",
      "AUUUTG\n",
      "AUUGTG\n",
      "AUUCTG\n",
      "AUGATG\n",
      "AUGUTG\n",
      "AUGGTG\n",
      "AUGCTG\n",
      "AUCATG\n",
      "AUCUTG\n",
      "AUCGTG\n",
      "AUCCTG\n",
      "AGAATG\n",
      "AGAUTG\n",
      "AGAGTG\n",
      "AGACTG\n",
      "AGUATG\n",
      "AGUUTG\n",
      "AGUGTG\n",
      "AGUCTG\n",
      "AGGATG\n",
      "AGGUTG\n",
      "AGGGTG\n",
      "AGGCTG\n",
      "AGCATG\n",
      "AGCUTG\n",
      "AGCGTG\n",
      "AGCCTG\n",
      "ACAATG\n",
      "ACAUTG\n",
      "ACAGTG\n",
      "ACACTG\n",
      "ACUATG\n",
      "ACUUTG\n",
      "ACUGTG\n",
      "ACUCTG\n",
      "ACGATG\n",
      "ACGUTG\n",
      "ACGGTG\n",
      "ACGCTG\n",
      "ACCATG\n",
      "ACCUTG\n",
      "ACCGTG\n",
      "ACCCTG\n",
      "ATAGGC\n",
      "ATUGGC\n",
      "ATGGGC\n",
      "ATCGGC\n",
      "ATAAGG\n",
      "ATAUGG\n",
      "ATAGGG\n",
      "ATACGG\n",
      "ATUAGG\n",
      "ATUUGG\n",
      "ATUGGG\n",
      "ATUCGG\n",
      "ATGAGG\n",
      "ATGUGG\n",
      "ATGGGG\n",
      "ATGCGG\n",
      "ATCAGG\n",
      "ATCUGG\n",
      "ATCGGG\n",
      "ATCCGG\n",
      "ATAAAG\n",
      "ATAAUG\n",
      "ATAAGG\n",
      "ATAACG\n",
      "ATAUAG\n",
      "ATAUUG\n",
      "ATAUGG\n",
      "ATAUCG\n",
      "ATAGAG\n",
      "ATAGUG\n",
      "ATAGGG\n",
      "ATAGCG\n",
      "ATACAG\n",
      "ATACUG\n",
      "ATACGG\n",
      "ATACCG\n",
      "ATUAAG\n",
      "ATUAUG\n",
      "ATUAGG\n",
      "ATUACG\n",
      "ATUUAG\n",
      "ATUUUG\n",
      "ATUUGG\n",
      "ATUUCG\n",
      "ATUGAG\n",
      "ATUGUG\n",
      "ATUGGG\n",
      "ATUGCG\n",
      "ATUCAG\n",
      "ATUCUG\n",
      "ATUCGG\n",
      "ATUCCG\n",
      "ATGAAG\n",
      "ATGAUG\n",
      "ATGAGG\n",
      "ATGACG\n",
      "ATGUAG\n",
      "ATGUUG\n",
      "ATGUGG\n",
      "ATGUCG\n",
      "ATGGAG\n",
      "ATGGUG\n",
      "ATGGGG\n",
      "ATGGCG\n",
      "ATGCAG\n",
      "ATGCUG\n",
      "ATGCGG\n",
      "ATGCCG\n",
      "ATCAAG\n",
      "ATCAUG\n",
      "ATCAGG\n",
      "ATCACG\n",
      "ATCUAG\n",
      "ATCUUG\n",
      "ATCUGG\n",
      "ATCUCG\n",
      "ATCGAG\n",
      "ATCGUG\n",
      "ATCGGG\n",
      "ATCGCG\n",
      "ATCCAG\n",
      "ATCCUG\n",
      "ATCCGG\n",
      "ATCCCG\n"
     ]
    }
   ],
   "source": [
    "print 4**6\n",
    "for j in check_insert('ATGGCC'):\n",
    "    print j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from pyevolve import GenomeBase\n",
    "from collections import Counter, defaultdict as ddict\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "class BranchRep:\n",
    "    \n",
    "    def __init__(self,part1, part2, btype):\n",
    "        self.part1 =  part1\n",
    "        self.part2 = part2\n",
    "        self.btype =  btype\n",
    "    \n",
    "    def get_hash(self, partition):\n",
    "        return hashlib.sha384(\",\".join(sorted(partition))).hexdigest()\n",
    "\n",
    "    def link_leaf(self):\n",
    "        return self.btype == 0\n",
    "        \n",
    "    def __hash__(self):\n",
    "        return self.get_hash(self.part1) + self.get_hash(self.part2)\n",
    "        \n",
    "        \n",
    "        \n",
    "class Utils:\n",
    "    \n",
    "    @staticmethod\n",
    "    def shuffle_map(gmap):\n",
    "        for k,v in gmap.items():\n",
    "            np.random.shuffle(v)\n",
    "        return gmap\n",
    "\n",
    "    @staticmethod\n",
    "    def initialize(genome, gmap):\n",
    "        gmap = Utils.shuffle_map(gmap)\n",
    "        pos = {}\n",
    "        for node in genome.tree:\n",
    "            node.add_features(species=node.name)\n",
    "            pos[node.species] = pos.get(node.species, 0)\n",
    "            node.name =  gmap[node.species][pos[node.species]]\n",
    "            pos[node.species] += 1\n",
    "    \n",
    "\n",
    "    @staticmethod\n",
    "    def iter_partition(tree):\n",
    "        find_root = False\n",
    "        for node in tree.iter_descendants(\"levelorder\"):\n",
    "            part1 = set(node.get_leaf_names())\n",
    "            part2 = set(tree.get_leaf_names()) - part1\n",
    "            nodes = [node]\n",
    "            importance = 1\n",
    "            if(not node.up.is_root()):\n",
    "                if node.is_leaf():\n",
    "                    importance = 0\n",
    "                yield part1, part2, nodes, importance\n",
    "            elif not find_root:\n",
    "                find_root = True\n",
    "                nodes.extend(node.get_sisters())\n",
    "                yield part1, part2, nodes, importance\n",
    "\n",
    "            \n",
    "    @staticmethod\n",
    "    def two_step_branch_selection(candbranch, prob):\n",
    "        btype = np.random.choice([1,0], p=[prob, 1-prob])\n",
    "        new_cands = [x for x in candbranch if x.btype == btype]\n",
    "        # if it's empty, then the list only contains abs(btype-1)\n",
    "        if not new_cands:\n",
    "            new_cands = candbranch      \n",
    "        selected_branch = np.random.choice(new_cands)\n",
    "            \n",
    "    \n",
    "    @staticmethod\n",
    "    def find_and_swap(branch, g1, g2):\n",
    "        \n",
    "        if not branch.link_leaf():\n",
    "            node1 = g1.tree.get_common_ancestor(branch.part1)\n",
    "            node2 = g1.tree.get_common_ancestor(branch.part2)\n",
    "            g1swap,  g1part = (node1,  branch.part1) if node2.is_root() else (node2, branch.part2)\n",
    "            g2swap = g2.tree.get_common_ancestor(g1part)\n",
    "            \n",
    "        else:\n",
    "            leaf_node = (branch.part1 if len(branch.part1) == 1 else branch.part2)[0]\n",
    "            g1swap = g1.tree&leaf_node\n",
    "            g2swap = g2.tree&leaf_node\n",
    "                    \n",
    "                \n",
    "        g1_parent = g1swap.up\n",
    "        g2_parent = g2swap.up\n",
    "        g2_parent.add_child(g1swap.detach())\n",
    "        g1_parent.add_child(g2swap.detach())\n",
    "        return g1, g2\n",
    "        \n",
    "        \n",
    "    @staticmethod\n",
    "    def permute_seq(genome, spec):\n",
    "        nlist = genome.tree.search_nodes(species=spec)\n",
    "        if len(nlist)>1:\n",
    "            n1, n2 = np.random.choice(nlist, 2, replace=False)\n",
    "            n1.name, n2.name = n2.name, n1.name            \n",
    "        \n",
    "    \n",
    "    @staticmethod\n",
    "    def crossover(genome, **args):\n",
    "        gdad = args['dad']\n",
    "        gmom = args['mom']\n",
    "        genome1 = gdad.clone()\n",
    "        genome2 = gmom.clone()\n",
    "        prob = max(genome1.intbrnp, genome2.intbrnp)\n",
    "        current_dict = ddict(int)\n",
    "        genomes = [genome1, genome2]\n",
    "        for g in genomes:\n",
    "            for part1, part2, nodes, imp in Utils.iter_partition(g.tree):\n",
    "                brnch = BranchRep(part1, part2, imp)\n",
    "                current_dict[brnch] += 1\n",
    "        \n",
    "        candidates = [k for k,v in current_dict.items() if v > 1]\n",
    "        selected_branch = Utils.two_step_branch_selection(candidates, prob)\n",
    "        return find_and_swap(selected_branch, genome1, genome2)\n",
    "        \n",
    "        \n",
    "    \n",
    "    @staticmethod\n",
    "    def evaluator(genome):\n",
    "        pass\n",
    "    \n",
    "        \n",
    "    @staticmethod\n",
    "    def mutate(genome, **args):\n",
    "        if args[\"pmut\"] <= 0.0:\n",
    "            return 0\n",
    "        nspec = genome.get_spec_len()\n",
    "        av_mutation =  args[\"pmut\"]*nspec\n",
    "        spec_list = genome.spcount.keys()\n",
    "        if av_mutation < 1.0:\n",
    "            av_mutation = 0\n",
    "            for spec in spec_list:\n",
    "                if np.random.rand() < args[\"pmut\"]:\n",
    "                    #mutate\n",
    "                    av_mutation += 1\n",
    "        else:\n",
    "            for i in range(int(round(av_mutation))):\n",
    "                # choose a random \n",
    "                spec = np.random.choice(spec_list)\n",
    "                # mutate\n",
    "                \n",
    "            \n",
    "        return av_mutation\n",
    "    \n",
    "    \n",
    "class GPolySolver(GenomeBase.GenomeBase):\n",
    "    def __init__(self, tree, intbrnp=0.95):\n",
    "        self.tree =  tree\n",
    "        self.intbrnp = intbrnp\n",
    "        self.spcount =  Counter(tree.get_leaf_names())\n",
    "        self.initializator.set(Utils.initialize)\n",
    "        self.mutator.set(Utils.mutator)\n",
    "        self.crossover.set(Utils.crossover)\n",
    "        self.evaluator.set(Utils.evaluator)\n",
    "\n",
    "    def get_spec_len(self):\n",
    "        return len(self.spcount.keys())\n",
    "                   \n",
    "    def evaluate(self, **args):\n",
    "        \"\"\" Called to evaluate genome\n",
    "        :param args: this parameters will be passes to the evaluator\n",
    "        \"\"\"\n",
    "        self.resetStats()\n",
    "        for it in self.evaluator.applyFunctions(self, **args):\n",
    "            self.score += it\n",
    "\n",
    "        \n",
    "    def initialize(self, **args):\n",
    "        \"\"\" Called to initialize genome\n",
    "        :param args: this parameters will be passed to the initializator\n",
    "        \"\"\"\n",
    "        for it in self.initializator.applyFunctions(self, **args):\n",
    "            pass\n",
    "    \n",
    "\n",
    "    def mutate(self, **args):\n",
    "        \"\"\" Called to mutate the genome\n",
    "        :param args: this parameters will be passed to the mutator\n",
    "        :rtype: the number of mutations returned by mutation operator\n",
    "        \"\"\"\n",
    "        nmuts = 0\n",
    "        for it in self.mutator.applyFunctions(self, **args):\n",
    "            nmuts += it\n",
    "        return nmuts\n",
    "\n",
    "    def copy(self, g):\n",
    "        \"\"\" Copy the current GenomeBase to 'g'\n",
    "        :param g: the destination genome\n",
    "        .. note:: If you are planning to create a new chromosome representation, you\n",
    "                **must** implement this method on your class.\n",
    "        \"\"\"\n",
    "        GenomeBase.copy(self, g)\n",
    "        g.tree =  self.tree.copy()\n",
    "        g.spcount = self.spcount\n",
    "        g.intbrnp = self.intbrnp\n",
    "        \n",
    "    def clone(self):\n",
    "        \"\"\" Clone this GenomeBase\n",
    "        :rtype: the clone genome\n",
    "        .. note:: If you are planning to create a new chromosome representation, you\n",
    "            **must** implement this method on your class.\n",
    "        \"\"\"\n",
    "        newcopy = GPolySolver(None)\n",
    "        self.copy(newcopy)\n",
    "        return newcopy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "         /-aaaaaaaaab\n",
      "      /-|\n",
      "     |  |   /-aaaaaaaaac\n",
      "   /-|   \\-|\n",
      "  |  |      \\-aaaaaaaaaa\n",
      "  |  |\n",
      "--|   \\-aaaaaaaaaf\n",
      "  |\n",
      "  |   /-aaaaaaaaad\n",
      "   \\-|\n",
      "      \\-aaaaaaaaae\n",
      "\n",
      "   /-aaaaaaaaaf\n",
      "--|\n",
      "  |   /-aaaaaaaaae\n",
      "   \\-|\n",
      "     |   /-aaaaaaaaad\n",
      "      \\-|\n",
      "        |   /-aaaaaaaaac\n",
      "         \\-|\n",
      "           |   /-aaaaaaaaab\n",
      "            \\-|\n",
      "               \\-aaaaaaaaaa\n"
     ]
    }
   ],
   "source": [
    "from ete3 import Tree\n",
    "t1 = Tree(\"(((aaaaaaaaab:1,(aaaaaaaaac:1,aaaaaaaaaa:1)1:1)1:1,aaaaaaaaaf:1)1:1,(aaaaaaaaad:1,aaaaaaaaae:1)1:1);\")\n",
    "t2 = Tree(\"(aaaaaaaaaf:1,(aaaaaaaaae:1,(aaaaaaaaad:1,(aaaaaaaaac:1,(aaaaaaaaab:1,aaaaaaaaaa:1)1:1)1:1)1:1)1:1);\")\n",
    "print t1\n",
    "print t2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "br1 = BranchRep(['aaaaaaaaaa', 'aaaaaaaaab','aaaaaaaaac'], ['aaaaaaaaae', 'aaaaaaaaaf', 'aaaaaaaaad'], 1)\n",
    "br2 = BranchRep(['aaaaaaaaaa'], ['aaaaaaaaac', 'aaaaaaaaad', 'aaaaaaaaab', 'aaaaaaaaae', 'aaaaaaaaaf'], 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def find_and_swap(branch, t1, t2):\n",
    "\n",
    "    if not branch.link_leaf():\n",
    "        node1 = t1.get_common_ancestor(branch.part1)\n",
    "        node2 = t1.get_common_ancestor(branch.part2)\n",
    "        g1swap,  g1part = (node1,  branch.part1) if node2.is_root() else (node2, branch.part2)\n",
    "        g2swap = t2.get_common_ancestor(g1part)\n",
    "\n",
    "    else:\n",
    "        leaf_node = (branch.part1 if len(branch.part1) == 1 else branch.part2)[0]\n",
    "        g1swap = t1&leaf_node\n",
    "        g2swap = t2&leaf_node\n",
    "        \n",
    "    g1_parent = g1swap.up\n",
    "    g2_parent = g2swap.up\n",
    "    g2_parent.add_child(g1swap.detach())\n",
    "    g1_parent.add_child(g2swap.detach())\n",
    "    return t1, t2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "swap don -------------\n",
      "\n",
      "         /-aaaaaaaaab\n",
      "      /-|\n",
      "     |  |   /-aaaaaaaaac\n",
      "   /-|   \\-|\n",
      "  |  |      \\-aaaaaaaaaa\n",
      "  |  |\n",
      "--|   \\-aaaaaaaaaf\n",
      "  |\n",
      "  |   /-aaaaaaaaad\n",
      "   \\-|\n",
      "      \\-aaaaaaaaae\n",
      "\n",
      "      /-aaaaaaaaaf\n",
      "   /-|\n",
      "  |  |   /-aaaaaaaaac\n",
      "  |   \\-|\n",
      "  |     |   /-aaaaaaaaab\n",
      "--|      \\-|\n",
      "  |         \\-aaaaaaaaaa\n",
      "  |\n",
      "  |   /-aaaaaaaaad\n",
      "   \\-|\n",
      "      \\-aaaaaaaaae\n",
      "-------------------------------------------\n",
      "\n",
      "   /-aaaaaaaaaf\n",
      "--|\n",
      "  |   /-aaaaaaaaae\n",
      "   \\-|\n",
      "     |   /-aaaaaaaaad\n",
      "      \\-|\n",
      "        |   /-aaaaaaaaac\n",
      "         \\-|\n",
      "           |   /-aaaaaaaaab\n",
      "            \\-|\n",
      "               \\-aaaaaaaaaa\n",
      "\n",
      "   /-aaaaaaaaaf\n",
      "--|\n",
      "  |   /-aaaaaaaaae\n",
      "   \\-|\n",
      "     |   /-aaaaaaaaad\n",
      "      \\-|\n",
      "        |   /-aaaaaaaaab\n",
      "         \\-|\n",
      "           |   /-aaaaaaaaac\n",
      "            \\-|\n",
      "               \\-aaaaaaaaaa\n"
     ]
    }
   ],
   "source": [
    "t11 = t1.copy()\n",
    "t111 = t1.copy()\n",
    "t22 = t2.copy()\n",
    "t222 = t2.copy()\n",
    "\n",
    "print \"swap don -------------\"\n",
    "r1, r2 = find_and_swap(br1, t11, t22)\n",
    "\n",
    "print t1\n",
    "print r1\n",
    "print '-------------------------------------------'\n",
    "print t2\n",
    "print r2\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Cannot take a larger sample than population when 'replace=False'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-141-35c615f7f8b9>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0mg\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchoice\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mg\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m2\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mreplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mmtrand.pyx\u001b[0m in \u001b[0;36mmtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:12035)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Cannot take a larger sample than population when 'replace=False'"
     ]
    }
   ],
   "source": [
    "g = range(1,2)\n",
    "np.random.choice(g, 2, replace=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
